File: leetcode/c++/394_decode_string.cpp
// https://leetcode.cn/problems/decode-string/description

#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
    string decodeString(string s) {
        string res;
        stack<int> stack_num;
        stack<string> stack_res;
        int num = 0;
        for (char c : s) {
            if (c == '[') {
                stack_num.push(num);
                stack_res.push(res);
                num = 0;
                res = "";
            } else if (c == ']') {
                string tmp;
                for (int i = stack_num.top(); i > 0; i--)
                    tmp += res;
                stack_num.pop();
                res = stack_res.top() + tmp;
                stack_res.pop();
            } else if (c <= '9' && c >='0')
                num = num * 10 + (c - '0');
            else 
                res += c;
        }
        return res;
    }
};

// 时间复杂度为O(n)，空间复杂度为o(n)

File: leetcode/golang/394_decode_string.go
func decodeString(s string) string {
    var res string
    var numStack []int
    var resStack []string
    num := 0
    for i := 0; i < len(s); i++ {
        if s[i] == '[' {
            numStack = append(numStack, num)
            resStack = append(resStack, res)
            num = 0
            res = ""
        } else if s[i] == ']' {
            tmp := ""
            for j := 0; j < numStack[len(numStack)-1]; j++ {
                tmp += res
            }
            numStack = numStack[: len(numStack)-1]
            res = resStack[len(resStack)-1] + tmp
            resStack = resStack[: len(resStack)-1]
        } else if s[i] >= '0' && s[i] <= '9' {
            num = num * 10 + int(s[i] - '0')
        } else {
            res += string(s[i])
        }
    }
    return res
}